ibazel/profiler/profiler.go

```diff
@@ -0,0 +1,429 @@
+	}
+	i.server.Addr = makeAddr(port)
+
+	// Handle profiler.js requests
+	router.HandleFunc("/profiler.js", jsHandler(i))
+
+	// Handle profiler events
+	router.HandleFunc("/profiler-event", profilerEventHandler(i))
+
+	// Create listener
+	l, err := net.Listen("tcp", makeAddr(port))
+	if err != nil {
+		i.closeServer()
+		return err
+	}
+
+	fmt.Fprintf(os.Stderr, "[profiler] listening on %s\n", i.server.Addr)
+	err = i.server.Serve(l)
+	i.closeServer()
+	return err
+}
+
+func (i *Profiler) startEvent(info *map[string]string) {
+	event := profileEvent{}
+	event.Type = "IBAZEL_START"
+	if info != nil {
+		event.IBazelVersion = i.version
+		event.BazelVersion = (*info)["release"]
+		event.MaxHeapSize = (*info)["max-heap-size"]
+		event.CommittedHeapSize = (*info)["committed-heap-size"]
+	}
+	i.lock.Lock()
+	i.processEvent(&event)
+	i.lock.Unlock()
+}
+
+func (i *Profiler) buildEvent(eventType string) {
+	i.lock.Lock()
+	i.buildingIteration()
+	event := profileEvent{}
+	event.Type = eventType
+	event.Changes = i.changes
+	i.processEvent(&event)
+	i.lock.Unlock()
+}
+
+func (i *Profiler) changeEvent(eventType string, change string) {
+	i.lock.Lock()
+	i.newIteration()
+	event := profileEvent{}
+	event.Type = eventType
+	event.Change = change
+	i.processEvent(&event)
+	i.changes = append(i.changes, change)
+	i.lock.Unlock()
+}
+
+func (i *Profiler) reloadTriggeredEvent() {
+	i.lock.Lock()
+	i.iterationReloadTriggered = true
+	event := profileEvent{}
+	event.Type = "RELOAD_TRIGGERED"
+	event.Changes = i.changes
+	i.processEvent(&event)
+	i.lock.Unlock()
+}
+
+func (i *Profiler) remoteEvent(remoteEvent *profilerRemoteEvent) {
+	i.lock.Lock()
+	if !i.iterationReloadTriggered {
+		fmt.Fprintf(os.Stderr, "Ignoring unexpected remote event\n")
+		return
+	}
+	event := profileEvent{}
+	event.Type = "REMOTE_EVENT"
+	event.RemoteType = remoteEvent.Type
+	event.RemoteTime = remoteEvent.Time
+	event.RemoteElapsed = remoteEvent.TimeSinceNavigationStart
+	event.RemoteData = remoteEvent.Data
+	i.processEvent(&event)
+	i.lock.Unlock()
+}
+
+func (i *Profiler) processEvent(event *profileEvent) {
+	if i.file != nil && event != nil {
+		// prepare the event
+		event.Iteration = i.iteration
+		event.Time = makeTimestamp()
+		event.Targets = i.targets
+		event.Elapsed = event.Time - i.iterationStartTime
+
+		// write the event to the output file
+		eventJson, _ := json.Marshal(event)
+		eventJson = append(eventJson, 10) // \n
+		_, err := i.file.Write(eventJson)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error writing to profile file: %v\n", err)
+		}
+	}
+}
+
+func (i *Profiler) newIteration() {
+	if i.iterationBuildStart {
+		i.iteration = randomString(16)
+		i.changes = make([]string, 0, 100)
+		i.iterationStartTime = makeTimestamp()
+		i.iterationBuildStart = false
+		i.iterationReloadTriggered = false
+	}
+}
+
+func (i *Profiler) buildingIteration() {
+	i.iterationBuildStart = true
+}
+
+func jsHandler(i *Profiler) http.HandlerFunc {
```

Comment:
Methods attached to objects hold their reference to the receiver. All the handlers you're writing can be attached to the (i *Profiler) without having to take it as a parameter. That means you don't need to nest the function in a function and you can have jsHandler directly implement the interface `func(ResponseWriter, *Request)`

I put together a partial demo that I hope makes this a little bit clearer https://play.golang.org/p/bVnlREy50U6